lib/elixir/lib/integer.ex

```diff
@@ -239,45 +239,24 @@ defmodule Integer do
     raise ArgumentError, "invalid base #{inspect base}"
   end
 
-  def parse(<<?-, rest::binary>>, base) do
-    case parse_digits(rest, base) do
-      {acc, bin} ->
-        {-acc, bin}
-      :error ->
-        :error
-    end
-  end
-
-  def parse(<<?+, rest::binary>>, base) do
-    parse_digits(rest, base)
-  end
-
-  def parse(<<rest::binary>>, base) do
-    parse_digits(rest, base)
-  end
+  def parse(<<?-, rest::binary>>, base), do: parse_digits(rest, base, "-")
+  def parse(<<?+, rest::binary>>, base), do: parse_digits(rest, base, "")
+  def parse(<<rest::binary>>, base), do: parse_digits(rest, base, "")
 
   digits = [{?0..?9, -?0}, {?A..?Z, 10 - ?A}, {?a..?z, 10 - ?a}]
 
   for {chars, diff} <- digits, char <- chars do
-    defp parse_digits(<<unquote(char), rest::binary>>, base)
-         when base > unquote(char + diff) do
-      parse_digits(rest, base, unquote(char + diff))
-    end
-  end
-
-  defp parse_digits(<<_::binary>>, _) do
-    :error
-  end
-
-  for {chars, diff} <- digits, char <- chars do
     defp parse_digits(<<unquote(char), rest::binary>>, base, acc)
          when base > unquote(char + diff) do
-      parse_digits(rest, base, base * acc + unquote(char + diff))
+      parse_digits(rest, base, <<acc::binary, unquote(char)>>)
     end
   end
 
-  defp parse_digits(<<rest::binary>>, _, acc) do
-    {acc, rest}
+  defp parse_digits(_, _, "-"), do: :error
```

Comment:
The compiler can optimise if the first argument in all clauses matches on a binary or single variable and returns the variable. This can be solved here, either by add `<<_::binary>>` as first argument or moving the `acc` matching inside the final body. You can see information about whats going on when compiling using `@compile :bin_opt_info` and more details at http://erlang.org/doc/efficiency_guide/binaryhandling.html#id65492.